sql/core/src/main/scala/org/apache/spark/sql/functions.scala

```diff
@@ -2590,8 +2590,21 @@ object functions {
    * @group datetime_funcs
    * @since 1.5.0
    */
-  def add_months(startDate: Column, numMonths: Int): Column = withExpr {
-    AddMonths(startDate.expr, Literal(numMonths))
+  def add_months(startDate: Column, numMonths: Int): Column = add_months(startDate, lit(numMonths))
+
+  /**
+   * Returns the date that is `numMonths` after `startDate`.
+   *
+   * @param startDate A date, timestamp or string. If a string, the data must be in a format that
+   *                  can be cast to a date, such as `yyyy-MM-dd` or `yyyy-MM-dd HH:mm:ss.SSSS`
+   * @param numMonths A column of the number of months to add to `startDate`, can be negative to
+   *                  subtract months
+   * @return A date, or null if `startDate` was a string that could not be cast to a date
+   * @group datetime_funcs
+   * @since 3.0.0
+   */
+  def add_months(startDate: Column, numMonths: Column): Column = withExpr {
```

Comment:
There was some discussion about it. I am sure there are more instances like this and we decided to don't add all overridden versions because this functions get too long. We already have multiple SQL functions that are not added into this file (see the comments on the top of this file)

Maybe we should only have the `Column` versions and deprecate/remove the others but it's kind of a big deal.

Reply:
So .. those have only been added when there are actual reasonable usecases.

Reply:
cc @cloud-fan FYI

Reply:
@HyukjinKwon Yes this is from a customer tickets. What about deprecate the other version function and only retain the column arg version ?

Reply:
other versions function we can provide implicit conversions from literal type to column.